### PR TITLE
Assignment5: Kevin

### PR DESCRIPTION
**Pull request by kflannery2001**
Worked solo

### CODE REVIEW

File: GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            this.Next = addedNode;
+            addedNode.Next = this;
+        }
+        // If there are more nodes in the list... I don't think I need this else statement
+        else
+        {
+            addedNode.Next = this.Next;
+            this.Next = addedNode;
+        }
```
Comment (Oscar-Ruelas) [inline GenericsHomework/Node.cs:46]: Wouldn't It had been better to use a while loop or a do-while inside the else statement for this part? I feel that you appending  a new node in between other nodes if the linked-list contains a large amount of nodes

File: GenericsHomework/Node.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    private bool Exists(T newData)
+    {
+        // Cycle through the list to check if the value already exists
+        var current = this;
```
Comment (Oscar-Ruelas) [inline GenericsHomework/Node.cs:52]: when you can remember  to specifically specify the Data Type you're using. I know to using var simplifies things but it helps other programmers know your intent.

File: GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            temp.Next = temp;
+        }
+
+        this.Next = this;
+    }
+
+    public override string ToString()
+    {
+        return "Data: " + Data;
+    }
```
Comment (Oscar-Ruelas) [inline GenericsHomework/Node.cs:83]: I feel that the ToString method could be improved like for example using "-->" to specify the order of the nodes .

**Review by Oscar-Ruelas — COMMENTED** [review]
Instructions

Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
The values should be homogeneous.
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and, therefore, may never be garbage collected. ✔
Create an Exists method to test whether a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔
Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌✔
Each circle contains n items and each item can belong to one or more Circle instances.
You are not required to use a Node from earlier in the homework for your Venn diagram implementation.
You are welcome to use existing BCL generic classes for the extra credit.
Implement System.Collections.Generic.ICollection<T> on the Node class ❌✔
Fundamentals

Place all shared project properties into a Directory.Build.Props file.
nullable reference types is enabled ❌✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌✔
and enabled .NET analyzers for both projects ❌✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔


Great Job Kevin 😄. You code looks great. I like how well commented your code is. Don't Forget to do the remaining part of the fundamentals.

File: GenericsHomework/Node.cs (hunk 2 above)
Comment (kflannery2001) [reply GenericsHomework/Node.cs:52]: Good call, fixed that

File: GenericsHomework/Node.cs (hunk 3 above)
Comment (kflannery2001) [reply GenericsHomework/Node.cs:83]: That's a good idea to provide more information about the whole linked list, thanks

File: GenericsHomework/Node.cs (hunk 1 above)
Comment (kflannery2001) [reply GenericsHomework/Node.cs:46]: That seems to make more sense, thanks

**Comment by Joshua-Lester3** [comment]
Hey Kevin, just a heads up that we require you to have a partner for these assignments. We deduct points for this.

**Review by quattro004 — APPROVED** [review]
Looks good

File: GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        
+    }
+
+    private bool Exists(T newData)
+    {
+        // Cycle through the list to check if the value already exists
+        Node<T> current = this;
+        do
+        {
+            if (current.Data!.Equals(newData))
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:52]: I don't see how you can guarantee that Data will not be null here

**Pull request closed without merging**